DRAFT: WSL-Ubuntu-Android Build Guide

D33r-Gee opened the pull request:
This is an updated version of PR #351
It has been rebased and the WSL Ubuntu Android install guide has been added to build-android.md
Let me know if there are any issues

D33r-Gee commented:
Rebased main, deleted WSL_APK_INSTALL.md and modified build-android.md so the wsl guide can be found there. Let me know if there's anything that needs to be updated @jarolrod

hebasto commented:
Please add PR description, update PR title, fix linter issues and clean commit history.

D33r-Gee commented:
> Please add PR description, update PR title, fix linter issues and clean commit history.

Took care of the above.
Re: clean commit history, please let me know if that's the way to do it or if you'd prefer a different approach?

pablomartin4btc commented:
Concept ACK

I'll try to validate the instructions following them and checking the built is correct. Is this exclusive to WSL or can be done on Ubuntu 22.04.

> Re: clean commit history, please let me know if that's the way to do it or if you'd prefer a different approach?

I think there's no need to have so many commits, you could squash them all into only one.

pablomartin4btc reviewed (commented):
Concept ACK

I'll try to validate the instructions following them and checking the built is correct. Is this exclusive to WSL or can be done on Ubuntu 22.04.

> Re: clean commit history, please let me know if that's the way to do it or if you'd prefer a different approach?

I think there's no need to have so many commits, you could squash them all into only one.

Also, perhaps you need to rebase?

D33r-Gee commented:
Thanks @pablomartin4btc for Concept Ack this PR

I will squash the commits and rebase early next week as I'm traveling for
the holiday.

On Wed, Nov 22, 2023, 20:33 pablomartin4btc ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> Concept ACK
>
> I'll try to validate the instructions following them and checking the
> built is correct. Is this exclusive to WSL or can be done on Ubuntu 22.04.
>
> Re: clean commit history, please let me know if that's the way to do it or
> if you'd prefer a different approach?
>
> I think there's no need to have so many commits, you could squash them all
> into only one.
>
> Also, perhaps you need to rebase?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin-core/gui-qml/pull/366#pullrequestreview-1745260072>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A2P6LNZVG3DCPSXE4EDXAPLYFZHQJAVCNFSM6AAAAAA4VFKKCSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTONBVGI3DAMBXGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

D33r-Gee commented:
> Is this exclusive to WSL or can be done on Ubuntu 22.04.

It should work with Ubuntu 22.04 proper.

Squashed the commits and rebased with `main`

Thanks again for reviewing @pablomartin4btc

pablomartin4btc reviewed (commented):
ACK a4bf253c4f7cc5125134b5d26e431677f12ffbaa

Sorry,  I think the 2nd commit ("Merge branch..." 9ca8bb2f92b13532e71bdbb340c3c041f0ca6288) is not necessary, only 1 commit.

I'll try to follow the guide at some point next week and give a deeper review. Thanks for working on this!

D33r-Gee commented:
thanks @pablomartin4btc, for taking a look... squashed it down to 1 commit.

> I'll try to follow the guide at some point next week and give a deeper review.

Sounds great!

> Thanks for working on this!

with pleasure

hebasto commented:
@D33r-Gee 

The diff looks too big (956 files) for a build guide, doesn't it?

D33r-Gee commented:
> The diff looks too big (956 files) for a build guide, doesn't it?

@hebasto, yes thanks for pointing it out. I didn't do proper house cleaning when synching my fork. It should be resolved now.

pablomartin4btc reviewed (commented):
tACK till **Step 4** included

I'll continue and finish it before the end of the next week. Looking forward to have this working and play with it.

pablomartin4btc reviewed (commented):
tACK till **Step 6** included, only 1 step left to finish...

I have some minor nits that I'll write them out when I complete the guide entirely.

pablomartin4btc commented on doc/build-android.md:
````diff
@@ -22,4 +22,215 @@ This is an example command for a default build with no disabled dependencies:
+    usbipd wsl list
+    ```
+
+3. Note down the `BUSID` of the device you want to connect to WSL. Then run the following command, replacing `<busid>` with the `BUSID` of your device:
````
Devices have to explicitly be exported using `usbip bind` before `usbipd` makes them available to other hosts.
So, before you run the `attach`command you need to bind the device: `usbipd bind -b <busid>`.

D33r-Gee replied:
good catch... it seems on my device the setting was already `SHARED`, however I'll add the `bind` suggestion in case the android device doesn't attach properly

pablomartin4btc reviewed (commented):
tACK 4523a0f2e854f2cd9de1968d127ec573c2c223ce

I've finished the entire guide, sorry it took me a while cos I've checked different devices and had to learn to use the tools.

I tried first with an old tablet (HUAWEI m3 lite 10") with Android 7.0 (API level #24) but the Bitcoin core app crashes due to some missing qt library (checked with `adb logcat` I needed to install "Ministro II" but it didn't fix it... I'll try later to build a qt bundle from Android Studio and send it to the device).

Secondly I've tried on these (friends and family - :smile: 🙏) phones: 

<details>
<summary> Google Pixel 6 (Android 14 - API Level 34)</summary>

![WhatsApp Image 2023-12-22 at 19 02 28](https://github.com/bitcoin-core/gui-qml/assets/110166421/33f26a52-a377-4d68-9ade-53675c34a65d)

![WhatsApp Image 2023-12-22 at 19 02 29](https://github.com/bitcoin-core/gui-qml/assets/110166421/3d84ea6c-ff4a-496c-8814-414c248e5fa5)

![WhatsApp Image 2023-12-22 at 19 02 29 (1)](https://github.com/bitcoin-core/gui-qml/assets/110166421/67e0fd3f-853e-40c3-a9e2-64a9d51d3689)

![WhatsApp Image 2023-12-22 at 19 02 30 (1)](https://github.com/bitcoin-core/gui-qml/assets/110166421/7503df66-7fd4-4f55-bc2a-41e799f80dae)

![WhatsApp Image 2023-12-22 at 19 02 30 (2)](https://github.com/bitcoin-core/gui-qml/assets/110166421/154d736e-e520-4bed-b766-c039d46d9362)

![WhatsApp Image 2023-12-22 at 19 02 29 (2)](https://github.com/bitcoin-core/gui-qml/assets/110166421/076cc400-859d-4967-8bbd-bb342adb5355)

</details>

<details>
<summary> Samsung A20 (Android 9 - API Level 28)</summary>

![WhatsApp Image 2023-12-28 at 12 21 30](https://github.com/bitcoin-core/gui-qml/assets/110166421/0e532434-f297-4579-8c80-191303775f40)

</details>

<details>
<summary> HUAWEI Y9 (Android 9 - API Level 28)</summary>

<img width="292" alt="image" src="https://github.com/bitcoin-core/gui-qml/assets/110166421/e7fe5df8-7a84-45a6-aae2-043153b3085a">


![WhatsApp Image 2023-12-23 at 00 34 19](https://github.com/bitcoin-core/gui-qml/assets/110166421/ad60b6a8-1c63-4ff8-9bfe-2847eed8f030)

![WhatsApp Image 2023-12-23 at 00 36 10](https://github.com/bitcoin-core/gui-qml/assets/110166421/023df763-7a6e-4c28-8656-e42e89228eac)

![WhatsApp Image 2023-12-23 at 00 48 06](https://github.com/bitcoin-core/gui-qml/assets/110166421/285515ab-5b71-43bd-a8a7-a2562c60ba4c)

![WhatsApp Image 2023-12-23 at 01 13 30](https://github.com/bitcoin-core/gui-qml/assets/110166421/7719f3c7-6cdb-4f2c-a9b0-ba0138bb6db7)

![WhatsApp Image 2023-12-23 at 01 52 40](https://github.com/bitcoin-core/gui-qml/assets/110166421/b1836f91-18f3-4811-86b9-1af30958db3f)

</details>

Finally, I've re-built it for my old smart watch...

<details>
<summary>HUAWEI watch 2 (32 bits, Android 8.0.0, API Level 26)</summary>


![Screenshot_20231227_223056](https://github.com/bitcoin-core/gui-qml/assets/110166421/94517fe2-a582-4e42-9bb0-472f0f0e8779)

![screenshot-modified](https://github.com/bitcoin-core/gui-qml/assets/110166421/0b354d09-c9bc-45f6-91a7-04cd819cee03)

![screenshot2-modified](https://github.com/bitcoin-core/gui-qml/assets/110166421/85d3085e-825c-4c2b-b77b-0c08009b5865)

![screenshot4-modified](https://github.com/bitcoin-core/gui-qml/assets/110166421/4a1eae16-86e5-4610-88de-3dd04ad084d2)

![screenshot5-modified](https://github.com/bitcoin-core/gui-qml/assets/110166421/ca053217-94af-4ea2-a78e-159613b8bcfc)

![screenshot6-modified](https://github.com/bitcoin-core/gui-qml/assets/110166421/74fb9827-e9fb-47c1-aa77-239baf3f66df)


![Screenshot_20231227_223641](https://github.com/bitcoin-core/gui-qml/assets/110166421/51ce128f-c871-4a1f-afce-ecd332d5d74a)

![Screenshot_20231227_223546](https://github.com/bitcoin-core/gui-qml/assets/110166421/89c93c7b-34ee-4900-96f0-e1d4bba5f26f)

![Screenshot_20231227_223355](https://github.com/bitcoin-core/gui-qml/assets/110166421/805636b5-45f5-4150-bdf0-153eec145d71)


</details>

I've noticed some issues regarding the user experience, layout, etc., on the different platforms/ devices related to QML itself so I'll leave that out of this PR which approach is only the WSL - Ubuntu Android Build Guide.

I think I'd leave this guide open to use both WSL and Ubuntu itself so I'm not sure you need to but in case you do, you can also install `usbipd `[there on Ubuntu](https://manpages.ubuntu.com/manpages/bionic/en/man1/usbipd.8.html).

D33r-Gee commented:
Hi @pablomartin4btc ,

Thanks so much for going through the whole guide and testing it on different devices,,, amazing!

> I've finished the entire guide, sorry it took me a while cos I've checked different devices and had to learn to use the tools.

No worries, happy you got through and thanks for suggesting improvements!

I will look them over, test them then update the commit in the coming days...

D33r-Gee commented on doc/build-android.md:
```diff
@@ -22,4 +22,215 @@ This is an example command for a default build with no disabled dependencies:
 
-After the depends are built configure with one of the resulting prefixes and run `make && make apk` in `src/qt`.
\ No newline at end of file
+After the depends are built configure with one of the resulting prefixes and run `make && make apk` in `src/qt`.
```
Great suggestion on the `configure` examples will add them...

RE: the grabble issue, I believe the key here is NOT to run `make` in the base directory but run it only in the `src/qt`, could you confirm if that works?

pablomartin4btc replied:
If I run `make` in `src/qt` I got this error:

```
  CXXLD    qt/bitcoin-qt
make[1]: *** No rule to make target 'test_bitcoin_qt'.  Stop.
make[1]: Leaving directory '/home/ubunpolk/src/bitcoin-core-gui-qml/src'
make: *** [Makefile:3: all] Error 2
```

That's why I run `make` in the base bitcoin directory, which works fine. Perhaps there's something wrong with my setup.
I'll try to do the Android build on a non-WSL Ubuntu and will let you know.

pablomartin4btc replied:
Both issues, `make` running under `src/qt` not working having to run it under bitcoin base project directory and `gradle` not working when running `make apk` having to touch `src/Makefile`, also happened when I followed and tested this guide on Ubuntu 22.04 (non-WSL).

D33r-Gee commented on doc/build-android.md:
````diff
@@ -22,4 +22,215 @@ This is an example command for a default build with no disabled dependencies:
+3. Note down the `BUSID` of the device you want to connect to WSL. Then run the following command, replacing `<busid>` with the `BUSID` of your device:
+
+    ```bash
+    usbipd wsl attach --busid <busid>
````
yep will update

D33r-Gee commented on doc/build-android.md:
````diff
@@ -22,4 +22,215 @@ This is an example command for a default build with no disabled dependencies:
+```
+
+More details on this step can be found [here](https://github.com/bitcoin/bitcoin/blob/master/doc/build-windows.md#compiling-with-windows-subsystem-for-linux).
+
````
At the top of this document the example given is:
    ANDROID_SDK=/home/user/Android/Sdk ANDROID_NDK=/home/user/Android/Sdk/ndk-bundle make HOST=aarch64-linux-android ANDROID_API_LEVEL=28 ANDROID_TOOLCHAIN_BIN=/home/user/Android/Sdk/ndk-bundle/toolchains/llvm/prebuilt/linux-x86_64/bin

    in your example, what do the `CC=...` and `CXX=...` add in terms building the proper depends?

pablomartin4btc replied:
I think it forces to build `depends` with the proper clang version for the required API Level.

D33r-Gee replied:
ah got it... ok so will leave the the extra `CC=` and `CXX=` from example to minimize confusion, however if you feel strongly about having them let me know?

pablomartin4btc replied:
I prefer to use them in all `make` calls to ensure the Android API Level for the particular architecture (arm64-v8a, armeabi-v7a, etc), but perhaps it's not required or I'm over-using them, maybe @hebasto can shed some light here.

D33r-Gee commented on doc/build-android.md:
````diff
@@ -22,4 +22,215 @@ This is an example command for a default build with no disabled dependencies:
+2. Once usbip is installed, list all USB devices connected to your PC by running the following command as Administrator in `cmd`:
+
+    ```bash
+    usbipd wsl list
````
Updating with proper command usbipd list

D33r-Gee commented:
@pablomartin4btc , thanks for your reviews and suggestions I've implemented them here https://github.com/bitcoin-core/gui-qml/pull/366/commits/8e20109af7cdd712bd951a58ced22716497b1cac
Let me know your thoughts and/or if there's anything I missed?

pablomartin4btc commented on doc/build-android.md:
```diff
@@ -22,4 +22,233 @@ This is an example command for a default build with no disabled dependencies:
+
+More details on this step can be found [here](https://github.com/bitcoin/bitcoin/blob/master/doc/build-windows.md#compiling-with-windows-subsystem-for-linux).
+
+Now, you can build the APKs using the guide found [here](https://github.com/bitcoin-core/gui-qml/blob/main/doc/build-android.md).
```
nit:

```suggestion
Now, you need to build the dependencies for the specific Android API Level, for which you can follow the guide found [here](https://github.com/bitcoin-core/gui-qml/blob/main/doc/build-android.md).
```

Maybe that guide above needs to be updated as it shows an example for "a default build with no disabled dependencies", and could display the `make -C depends` suggested previously. And later below you also specify run `make && make apk` from `src/qt` as on the build-android guide.

In my opinion, I'd be more clear on the sub-steps here in Step 7 in order to build the APK and transfer/ install it to the device (including for each the other commands to make it work as the `qlogging.cpp` update, paths, etc):
1. build dependencies
2. configure with proper host
3. make
4. make apk
5. adb install

D33r-Gee replied:
Great idea on the having sub-steps

D33r-Gee commented:
new commit 7687e2cb347772a96f20039abe65f5d241e2e57f

Added your suggestions @pablomartin4btc

pablomartin4btc commented on doc/build-android.md:
```diff
@@ -22,4 +22,249 @@ This is an example command for a default build with no disabled dependencies:
+
+    More details on this step can be found [here](https://github.com/bitcoin/bitcoin/blob/master/doc/build-windows.md#compiling-with-windows-subsystem-for-linux).
+
+1. Now, you need to build the dependencies for the specific Android API Level, for which you can follow the guide found at the top of this doc [here](https://github.com/bitcoin-core/gui-qml/blob/main/doc/build-android.md).
```
```suggestion
1. Now, you need to build the [dependencies](https://github.com/bitcoin/bitcoin/blob/master/depends/README.md) for the specific Android API Level and architecture of your Android device (as you discovered in #Step 5 - point 6), for which you can follow the guide found at the top of this doc [here](https://github.com/bitcoin-core/gui-qml/blob/main/doc/build-android.md).
```

pablomartin4btc reviewed (approved):
tACK 7687e2cb347772a96f20039abe65f5d241e2e57f

Tested on Ubuntu 22.04 (non-WSL), so I think we can use this guide for both, perhaps you need to identify and separate the steps specifically for WSL (including usage of `usbip`, which is not required for non-WSL) and that's all I can say, it worked fine for me.

pablomartin4btc commented:
> I think we can use this guide for both, perhaps you need to identify and separate the steps specifically for WSL (including usage of `usbip`, which is not required for non-WSL) and that's all I can say, it worked fine for me.

Sorry, what I mentioned above perhaps can be done on a follow-up, apologies if you started doing that already :pray:.

D33r-Gee commented:
> Sorry, what I mentioned above perhaps can be done on a follow-up, apologies if you started doing that already 🙏.

No worries... Apologies for the late response was focused on the AssumeUTXO integration

> I think we can use this guide for both...

That's a great idea, once I implement your latest suggestions perhaps it can be merged and we can open a new PR detailing both OSes or we can keep collaborating on this one till we feel it's ready to be merged...

@pablomartin4btc thoughts?

D33r-Gee committed:
doc: WSL-Ubuntu-Android Install Guide

D33r-Gee commented:
updated and rebased https://github.com/bitcoin-core/gui-qml/pull/366/commits/5b7da4aed5da0fe3a9b1ab2c5c87c74527c028a6

Thanks @pablomartin4btc for testing and the suggestions

pablomartin4btc reviewed (approved):
re ACK 5b7da4aed5da0fe3a9b1ab2c5c87c74527c028a6

> perhaps it can be merged and we can open a new PR detailing both OSes or we can keep collaborating on this one till we feel it's ready to be merged...

IMO this PR can be merge as it is, and we can open a new one to incorporate Ubuntu  
with some tweaks.

hebasto commented:
Building for Android is currently [broken](https://github.com/bitcoin/bitcoin/issues/29360) in the Bitcoin Core master branch. This is the main reason why there have not been syncing to the master branch recently.

I suggest to draft / postpone this PR until CMake + Qt6 are landed to the master branch.

D33r-Gee commented:
> I suggest to draft / postpone this PR until CMake + Qt6 are landed to the master branch.

Ah great to know... yes setting this PR to draft then

hebasto commented:
Closing for now.

Feel free to reopen this based on the new [default](https://github.com/bitcoin-core/gui-qml/tree/qt6) branch.
